frontend/app/src/util/AppNavigation.ts

```diff
@@ -140,6 +149,128 @@ export class V1Strategy {
+
+  constructor(parent: AppNavigation) {
+    this.parent = parent
+    this.mainScriptHash = null
+    this.appPages = []
+    this.mainPage = null
+    this.hideSidebarNav = null
+  }
+
+  handleNewSession(newSession: NewSession): MaybeStateUpdate {
+    this.mainScriptHash = newSession.mainScriptHash
+    this.hideSidebarNav = newSession.config?.hideSidebarNav ?? false
+
+    // We do not know the page name, so use an empty string version
+    document.title = " · Streamlit"
+
+    return [{ hideSidebarNav: this.hideSidebarNav }, () => {}]
+  }
+
+  handlePagesChanged(_pagesChangedMsg: PagesChanged): MaybeStateUpdate {
+    // This message does not apply to V2
+    return undefined
+  }
+
+  handlePageNotFound(pageNotFound: PageNotFound): MaybeStateUpdate {
+    const { pageName } = pageNotFound
+    this.parent.onPageNotFound(pageName)
+
+    return [
+      { currentPageScriptHash: this.mainScriptHash ?? "" },
+      () => {
+        this.parent.hostCommunicationMgr.sendMessageToHost({
+          type: "SET_CURRENT_PAGE_NAME",
+          currentPageName: "",
+          currentPageScriptHash: this.mainScriptHash ?? "",
+        })
+      },
+    ]
+  }
+
+  handleNavigation(navigationMsg: Navigation): MaybeStateUpdate {
+    const { sections, position, appPages } = navigationMsg
+
+    this.appPages = appPages
+    this.hideSidebarNav = this.hideSidebarNav || position === "hidden"
+
+    const currentPage = appPages.find(
+      p => p.pageScriptHash === navigationMsg.pageScriptHash
+    ) as IAppPage
+    const mainPage = appPages.find(p => p.isDefault) as IAppPage
+    this.mainPage = mainPage
+    const currentPageScriptHash = currentPage.pageScriptHash as string
+    const currentPageName = currentPage.isDefault
+      ? ""
+      : (currentPage.pageName as string).replaceAll(" ", "_")
+
+    document.title = `${currentPage.pageName as string} · Streamlit`
+    this.parent.onUpdatePageUrl(
+      mainPage.pageName as string,
+      currentPageName,
+      currentPage.isDefault ?? false
+    )
+
+    return [
+      {
+        appPages,
+        navSections: sections,
+        hideSidebarNav: this.hideSidebarNav,
+        currentPageScriptHash,
+      },
+      () => {
+        this.parent.hostCommunicationMgr.sendMessageToHost({
+          type: "SET_APP_PAGES",
+          appPages,
+        })
+
+        this.parent.hostCommunicationMgr.sendMessageToHost({
+          type: "SET_CURRENT_PAGE_NAME",
+          currentPageName: currentPageName,
+          currentPageScriptHash,
+        })
+      },
+    ]
+  }
+
+  findPageByUrlPath(pathname: string): IAppPage {
+    return (
+      this.appPages.find(appPage =>
+        // The page name is embedded at the end of the URL path, and if not, we are in the main page.
+        // See https://github.com/streamlit/streamlit/blob/1.19.0/frontend/src/App.tsx#L740
+        pathname.endsWith("/" + appPage.urlPathname)
+      ) ?? (this.mainPage as IAppPage)
+    )
+  }
+
+  clearPageElements(
```

Comment:
a more telling name might be `retainMainScriptElements` or `filterMainScriptElements` if I understood the logic correctly.